src/network/networkd-ndisc.c

```diff
@@ -824,6 +812,74 @@ static int ndisc_router_process_options(Link *link, sd_ndisc_router *rt) {
         }
 }
 
+static int ndisc_drop_outdated(Link *link, sd_ndisc_router *rt) {
```

Comment:
Is this function really needed? We already drop expired DNS entries in `ndisc_vacuum()`, and we pass the lifetimes of routes and addresses down to the kernel, so shouldn't it handle dropping those for us?

Is the issue that our local state gets out of sync with the kernel's in that case? If so, shouldn't we be dropping our own record of addresses/routes as soon as they expire instead of waiting around for an RA to happen to arrive? There's no guarantee we'll get an RA in a timely manner, and until one arrives we'll still be desynced.

Reply:
`ndisc_vacuum()` is only called when updating state files, and state files are updated only when `link_dirty()` is called.
If a RA is received without DNS related options, then without this change, we will not set the link dirty, and the state files will not be updated, even if some of the previously received DNS entries are outdated. Hence, we need to call it here.

About addresses, IIRC, the kernel deprecates an address if its lifetime becomes zero, but does not remove it. Hence, we need to remove it.

About routes, the lifetime setting is slightly new. We need to support old kernels, hence we need to remove outdated routes.

> There's no guarantee we'll get an RA in a timely manner, and until one arrives we'll still be desynced.

Yeah, we should introduce timer event source. Will update.